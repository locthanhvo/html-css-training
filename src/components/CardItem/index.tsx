import { DragEvent, memo, useCallback, useMemo } from 'react';
import {
  Box,
  Card,
  CardBody,
  Flex,
  Image,
  Stack,
  Text,
} from '@chakra-ui/react';

// Utils
import { calculateDaysLeft, countLines } from '@/utils';

// Icons
import {
  ClockIcon,
  AttachmentIcon,
  HamburgerMenuIcon,
} from '@/components/common/Icons';

interface CardItemProps {
  id: string;
  title?: string;
  description?: string;
  label?: { name: string; value: string }[];
  startDate?: string;
  endDate?: string;
  images?: string[];
  members?: { image: string; value: string }[];
  onClick?: (id: string) => void;
  onDragStart: (event: DragEvent<HTMLDivElement>, taskId: string) => void;
}

const CardItem = ({
  id,
  title,
  description,
  label,
  startDate,
  endDate,
  members,
  images,
  onClick,
  onDragStart,
}: CardItemProps) => {
  const daysLeft =
    startDate && endDate && calculateDaysLeft(startDate, endDate);

  const countLineDescription = description && countLines(description);

  const handleClickTask = useCallback(() => {
    onClick?.(id);
  }, [onClick, id]);

  const labelElements = useMemo(
    () =>
      Boolean(label?.length) && (
        <Flex gap={1}>
          {label?.map(({ name, value }) => (
            <Box key={`label-${value}`}>
              <Text size="sm" color="coolGray">
                {name}
              </Text>
            </Box>
          ))}
        </Flex>
      ),
    [label],
  );

  const memberElements = useMemo(
    () =>
      Boolean(members?.length) && (
        <Flex>
          {members?.map(({ image, value }) => (
            <Image
              borderRadius="full"
              boxSize="20px"
              src={image}
              alt="Avatar"
              key={value}
            />
          ))}
        </Flex>
      ),
    [members],
  );

  const handleDragStart = useCallback(
    (event: DragEvent<HTMLDivElement>) => {
      onDragStart(event, id);
    },
    [onDragStart],
  );

  return (
    <Card
      data-testid="cart-item"
      cursor="pointer"
      w={230}
      borderRadius="md"
      draggable
      onClick={handleClickTask}
      _hover={{
        border: '2px solid',
        borderColor: 'lightRed',
      }}
      onDragStart={handleDragStart}
    >
      <CardBody p={0}>
        <Image
          display={images?.length ? 'block' : 'none'}
          borderTopRadius="md"
          src={images?.[0]}
          alt={`${title} image`}
          w={230}
          h={129}
        />

        <Stack spacing="3" p={3}>
          <Flex
            gap={2}
            direction={(label || []).length >= 2 ? 'column' : 'row'}
            justifyContent="space-between"
          >
            {labelElements}

            <Flex>{memberElements}</Flex>
          </Flex>
          <Text size="sm" color="primary">
            {title}
          </Text>

          <Flex justifyContent="space-between" alignItems="center">
            <Flex
              display={Number(countLineDescription) > 1 ? 'flex' : 'none'}
              justifyContent="space-between"
              alignItems="center"
            >
              <HamburgerMenuIcon />
              <Text textAlign="center" size="sm" color="coolGray">
                {countLineDescription}
              </Text>
            </Flex>

            <Flex
              display={images?.length ? 'flex' : 'none'}
              justifyContent="space-between"
              alignItems="center"
            >
              <AttachmentIcon />
              <Text size="sm" color="coolGray">
                {images?.length}
              </Text>
            </Flex>

            <Flex
              display={daysLeft ? 'flex' : 'none'}
              justifyContent="space-between"
              alignItems="center"
            >
              <ClockIcon {...(Number(daysLeft) === 1 && { color: 'red' })} />
              <Text size="sm" color={Number(daysLeft) > 1 ? 'coolGray' : 'red'}>
                {daysLeft} Days Left
              </Text>
            </Flex>
          </Flex>
        </Stack>
      </CardBody>
    </Card>
  );
};

export default memo(CardItem);
