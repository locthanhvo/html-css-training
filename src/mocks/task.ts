import AvatarUser from '@/assets/images/avatar-small-1.png';
import AvatarUser2 from '@/assets/images/avatar-small-2.png';
import AvatarUser3 from '@/assets/images/avatar-small-3.png';

export const TASK_DETAIL = {
  id: '1',
  title: 'Global Resorts Network',
  label: [
    {
      name: 'Space Tasks 1',
      value: 'space-tasks-1',
    },
  ],
  members: [
    { value: '1', name: 'Member 1', image: AvatarUser },
    { value: '2', name: 'Member 2', image: AvatarUser2 },
    { value: '3', name: 'Member 3', image: AvatarUser3 },
  ],
  created_at: '2022-01-01',
  updated_at: '2022-01-01',
  startDate: '2024-07-01T00:00:00.000Z',
  endDate: '2024-07-03T00:00:00.000Z',
  boardId: '93d1f6a8-c312-436a-886f-a646bf424209',
};

export const TASK_DETAIL_2 = {
  id: '2',
  title: 'Global Suns Network',
  label: [
    {
      name: 'Space Tasks 1',
      value: 'space-tasks-1',
    },
    {
      name: 'Space Tasks 2',
      value: 'space-tasks-2',
    },
  ],
  members: [
    { value: '1', name: 'Member 1', image: AvatarUser },
    { value: '2', name: 'Member 2', image: AvatarUser2 },
    { value: '3', name: 'Member 3', image: AvatarUser3 },
  ],
  images: ['https://images.unsplash.com/photo-1555041469-a586c61ea9bc'],
  description: '<p>hello</p><p>hi</p>',
  startDate: '2024-07-01T00:00:00.000Z',
  endDate: '2024-07-02T00:00:00.000Z',
  created_at: '2022-01-01',
  updated_at: '2022-01-01',
  boardId: '93d1f6a8-c312-436a-886f-a646bf424209',
};

export const LIST_TASK = {
  id: '1',
  title: 'to do',
  total: 2,
  color: 'electricPurple',
  tasks: [TASK_DETAIL, TASK_DETAIL_2],
};

export const MOCK_TASKS = [
  {
    id: '550e8400-e29b-41d4-a716-446655440000',
    label: [
      {
        name: 'Space Tasks 1',
        value: 'space-tasks-1',
      },
    ],
    title: 'Global Suns Network',
    description: 'description 2 \n description 3',
    images: ['https://images.unsplash.com/photo-1555041469-a586c61ea9bc'],
    startDate: '2024-07-01T00:00:00.000Z',
    endDate: '2024-07-10T00:00:00.000Z',
    boardId: '93d1f6a8-c312-436a-886f-a646bf424209',
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440001',
    label: [
      {
        name: 'Space Tasks 1',
        value: 'space-tasks-1',
      },
    ],
    title: 'Make Money Online Though',
    boardId: '93d1f6a8-c312-436a-886f-a646bf424209',
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440002',
    label: [
      {
        name: 'Space Tasks 1',
        value: 'space-tasks-1',
      },
    ],
    title: 'Characteristics Of A Successful',
    description: 'Task 3 description',
    boardId: '93d1f6a8-c312-436a-886f-a646bf424209',
  },
];

export const MOCK_BOARDS = [
  {
    id: '93d1f6a8-c312-436a-886f-a646bf424209',
    title: 'Todo',
    color: 'electricPurple',
  },
  {
    id: '93d1f6a8-c312-436a-886f-a646bf424208',
    title: 'In Work',
    color: 'skyBlue',
  },
  {
    id: '93d1f6a8-c312-436a-886f-a646bf424207',
    title: 'Review',
    color: 'yellow',
  },
  {
    id: '9b1a1aa0-afc0-4bdb-90f7-7d8330ec254b',
    title: 'Done',
    color: 'green',
  },
];
