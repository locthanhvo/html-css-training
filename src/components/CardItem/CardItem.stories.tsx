import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import CardItem from '.';

// Mocks
import { MOCK_TASK_DETAIL_FIRST, MOCK_TASK_DETAIL_SECOND } from '@/mocks';

const { id, title, images, description, label, startDate, endDate, members } =
  MOCK_TASK_DETAIL_SECOND;

export default {
  title: 'Components/CardItem',
  component: CardItem,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
} as Meta;

type Story = StoryObj<typeof CardItem>;

export const Default: Story = {
  args: {
    id: MOCK_TASK_DETAIL_FIRST.id,
    title: MOCK_TASK_DETAIL_FIRST.title,
    onClick: () => {},
  },
};

export const FullDetail: Story = {
  args: {
    id: id,
    title: title,
    images: images,
    description: description,
    label: label,
    startDate: startDate,
    endDate: endDate,
    members: members,
    onClick: () => {},
  },
};
