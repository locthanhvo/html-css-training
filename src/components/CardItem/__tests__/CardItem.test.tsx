import { fireEvent, render, RenderResult } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import CardItem from '..';

// Mocks
import { MOCK_TASK_DETAIL_FIRST, MOCK_TASK_DETAIL_SECOND } from '@/mocks';

const mockOnDragStart = jest.fn();
const mockOnClick = jest.fn();

describe('CardItem Component', () => {
  let renderResult: RenderResult;
  const { id, title, images, members, label, startDate, endDate, description } =
    MOCK_TASK_DETAIL_SECOND;

  beforeEach(() => {
    renderResult = render(
      <BrowserRouter>
        <CardItem
          id={id}
          title={title}
          images={images}
          members={members}
          label={label}
          endDate={endDate}
          startDate={startDate}
          description={description}
          onDragStart={mockOnDragStart}
          onClick={mockOnClick}
        />
      </BrowserRouter>,
    );
  });

  it('should match snapshot', () => {
    const { id, title, endDate, startDate } = MOCK_TASK_DETAIL_FIRST;
    const { container } = render(
      <BrowserRouter>
        <CardItem
          id={id}
          title={title}
          endDate={endDate}
          startDate={startDate}
          onDragStart={mockOnDragStart}
        />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should call onClick', () => {
    const { getByText } = renderResult;

    getByText('Global Suns Network').click();
    expect(mockOnClick).toHaveBeenCalled();
  });

  it('should call onDragStart', () => {
    const { getByText } = renderResult;

    fireEvent.dragStart(getByText('Global Suns Network'));

    expect(mockOnDragStart).toHaveBeenCalled();
  });
});
