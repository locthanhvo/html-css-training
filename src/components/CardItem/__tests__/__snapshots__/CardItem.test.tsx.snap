// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem Component should match snapshot 1`] = `
<div>
  <div
    class="chakra-card css-12c38m6"
    data-testid="cart-item"
    draggable="true"
  >
    <div
      class="chakra-card__body css-old1by"
    >
      <img
        alt="Global Resorts Network image"
        class="chakra-image css-1pv0iot"
      />
      <div
        class="chakra-stack css-1kgz9m7"
      >
        <div
          class="css-4rk6vg"
        >
          <div
            class="css-k008qs"
          />
        </div>
        <p
          class="chakra-text css-1p9c8n6"
        >
          Global Resorts Network
        </p>
        <div
          class="css-69i1ev"
        >
          <div
            class="css-lcy17d"
          >
            <svg
              fill="none"
              height="31"
              viewBox="0 0 30 31"
              width="30"
            >
              <path
                clip-rule="evenodd"
                d="M7 12.5C7.55 12.5 8 12.05 8 11.5C8 10.95 7.55 10.5 7 10.5C6.45 10.5 6 10.95 6 11.5C6 12.05 6.45 12.5 7 12.5ZM8 15.5C8 16.05 7.55 16.5 7 16.5C6.45 16.5 6 16.05 6 15.5C6 14.95 6.45 14.5 7 14.5C7.55 14.5 8 14.95 8 15.5ZM8 19.5C8 20.05 7.55 20.5 7 20.5C6.45 20.5 6 20.05 6 19.5C6 18.95 6.45 18.5 7 18.5C7.55 18.5 8 18.95 8 19.5ZM24 15.5C24 16.05 23.55 16.5 23 16.5H11C10.45 16.5 10 16.05 10 15.5C10 14.95 10.45 14.5 11 14.5H23C23.55 14.5 24 14.95 24 15.5ZM23 20.5C23.55 20.5 24 20.05 24 19.5C24 18.95 23.55 18.5 23 18.5H11C10.45 18.5 10 18.95 10 19.5C10 20.05 10.45 20.5 11 20.5H23ZM11 12.5C10.45 12.5 10 12.05 10 11.5C10 10.95 10.45 10.5 11 10.5H23C23.55 10.5 24 10.95 24 11.5C24 12.05 23.55 12.5 23 12.5H11Z"
                fill="#C3CAD9"
                fill-rule="evenodd"
              />
            </svg>
            <p
              class="chakra-text css-8lo94t"
            />
          </div>
          <div
            class="css-lcy17d"
          >
            <svg
              fill="none"
              height="31"
              viewBox="0 0 30 31"
              width="30"
            >
              <path
                clip-rule="evenodd"
                d="M20.2539 20H9.67393C7.58393 20 5.72393 18.47 5.52393 16.39C5.29393 14.01 7.16393 12 9.50393 12H21.8639C23.1739 12 24.3639 12.94 24.4939 14.24C24.6439 15.74 23.4739 17 22.0039 17H11.5039C10.9539 17 10.5039 16.55 10.5039 16C10.5039 15.45 10.9539 15 11.5039 15H20.2539C20.6639 15 21.0039 14.66 21.0039 14.25C21.0039 13.84 20.6639 13.5 20.2539 13.5H11.6439C10.3339 13.5 9.14393 14.44 9.01393 15.74C8.86393 17.24 10.0339 18.5 11.5039 18.5H21.8339C23.9239 18.5 25.7839 16.97 25.9839 14.89C26.2139 12.5 24.3439 10.5 22.0039 10.5H9.73393C6.86393 10.5 4.29393 12.6 4.02393 15.46C3.72393 18.75 6.28393 21.5 9.50393 21.5H20.2539C20.6639 21.5 21.0039 21.16 21.0039 20.75C21.0039 20.34 20.6639 20 20.2539 20Z"
                fill="#C3CAD9"
                fill-rule="evenodd"
              />
            </svg>
            <p
              class="chakra-text css-1hflwzw"
            />
          </div>
          <div
            class="css-69i1ev"
          >
            <svg
              fill="none"
              height="31"
              viewBox="0 0 30 31"
              width="30"
            >
              <path
                clip-rule="evenodd"
                d="M18.4064 18.5314L15.0364 16.5314V12.0014C15.0364 11.6014 14.7164 11.2814 14.3164 11.2814H14.2564C13.8564 11.2814 13.5364 11.6014 13.5364 12.0014V16.7214C13.5364 17.0714 13.7164 17.4014 14.0264 17.5814L17.6764 19.7714C18.0164 19.9714 18.4564 19.8714 18.6564 19.5314C18.8664 19.1814 18.7564 18.7314 18.4064 18.5314ZM23.7164 8.29139L20.6364 5.73139C20.2164 5.38139 19.5864 5.43139 19.2264 5.86139C18.8764 6.28139 18.9364 6.91139 19.3564 7.27139L22.4264 9.83139C22.8464 10.1814 23.4764 10.1314 23.8364 9.70139C24.1964 9.28139 24.1364 8.65139 23.7164 8.29139ZM6.63638 9.83139L9.70638 7.27139C10.1364 6.91139 10.1964 6.28139 9.83638 5.86139C9.48638 5.43139 8.85638 5.38139 8.43638 5.73139L5.35638 8.29139C4.93638 8.65139 4.87638 9.28139 5.23638 9.70139C5.58638 10.1314 6.21638 10.1814 6.63638 9.83139ZM14.5364 7.28139C9.56638 7.28139 5.53638 11.3114 5.53638 16.2814C5.53638 21.2514 9.56638 25.2814 14.5364 25.2814C19.5064 25.2814 23.5364 21.2514 23.5364 16.2814C23.5364 11.3114 19.5064 7.28139 14.5364 7.28139ZM14.5364 23.2814C10.6764 23.2814 7.53638 20.1414 7.53638 16.2814C7.53638 12.4214 10.6764 9.28139 14.5364 9.28139C18.3964 9.28139 21.5364 12.4214 21.5364 16.2814C21.5364 20.1414 18.3964 23.2814 14.5364 23.2814Z"
                fill="#C3CAD9"
                fill-rule="evenodd"
              />
            </svg>
            <p
              class="chakra-text css-1hflwzw"
            >
              2
               Days Left
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
